File: claude-infer/src/panorama.ts
import axios from "axios";
import * as fs from "fs";
import * as path from "path";
import sharp from "sharp";

const STREET_VIEW_METADATA_URL = "https://maps.googleapis.com/maps/api/streetview/metadata";
const TMP_DIR = path.join(__dirname, "..", "tmp");
const PANORAMA_DIR = path.join(TMP_DIR, "panorama");

// パノラマタイルのベースURL（APIキー不要）
const TILE_BASE_URL = "https://streetviewpixels-pa.googleapis.com/v1/tile";

interface StreetViewMetadata {
  status: string;
  pano_id?: string;
  location?: { lat: number; lng: number };
  date?: string;
  copyright?: string;
}

interface PanoramaInfo {
  panoId: string;
  headingOffset: number; // パノラマのヘディングオフセット（度）
}

// ディレクトリを作成
function ensureDir(dir: string): void {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

// Street View メタデータを取得してpano_idを返す
export async function getPanoramaId(
  lat: number,
  lng: number,
  apiKey: string
): Promise<string> {
  const response = await axios.get<StreetViewMetadata>(STREET_VIEW_METADATA_URL, {
    params: {
      location: `${lat},${lng}`,
      key: apiKey,
      source: "outdoor",
    },
  });

  if (response.data.status !== "OK" || !response.data.pano_id) {
    throw new Error(`Street View not available: ${response.data.status}`);
  }

  if (response.data.copyright && !response.data.copyright.includes("Google")) {
    throw new Error(`User-contributed image skipped: ${response.data.copyright}`);
  }

  return response.data.pano_id;
}

// パノラマのヘディングオフセットを取得
async function getPanoramaHeadingOffset(panoId: string): Promise<number> {
  try {
    const url = `https://www.google.com/maps/photometa/v1?authuser=0&hl=ja&gl=jp&pb=!1m4!1smaps_sv.tactile!11m2!2m1!1b1!2m2!1sja!2sjp!3m3!1m2!1e2!2s${panoId}!4m57!1e1!1e2!1e3!1e4!1e5!1e6!1e8!1e12!2m1!1e1!4m1!1i48!5m1!1e1!5m1!1e2!6m1!1e1!6m1!1e2!9m36!1m3!1e2!2b1!3e2!1m3!1e2!2b0!3e3!1m3!1e3!2b1!3e2!1m3!1e3!2b0!3e3!1m3!1e8!2b0!3e3!1m3!1e1!2b0!3e3!1m3!1e4!2b0!3e3!1m3!1e10!2b1!3e2!1m3!1e10!2b0!3e3`;

    const response = await axios.get(url, {
      headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      },
    });

    // レスポンスから heading を抽出
    // フォーマット: )]}' の後にJSONが続く
    const text = response.data as string;
    const jsonStr = text.replace(/^\)\]\}'\n/, "");
    const data = JSON.parse(jsonStr);

    // data[1][0][5][0][1][2][0] にheadingがある
    // 構造: [[],[[[1],[2,"panoId"],...,[[[1],[[null,null,lat,lng],[...],[heading,pitch,roll],...]]]]]]
    const headingData = data?.[1]?.[0]?.[5]?.[0]?.[1]?.[2];
    if (headingData && Array.isArray(headingData) && headingData.length >= 1) {
      const heading = headingData[0];
      if (typeof heading === "number") {
        return heading;
      }
    }

    // デフォルトは0（オフセットなし）
    console.warn(`  警告: パノラマ ${panoId} のヘディングオフセットを取得できませんでした`);
    return 0;
  } catch (error) {
    console.warn(`  警告: photometa取得エラー:`, error instanceof Error ? error.message : error);
    return 0;
  }
}

// パノラマタイルを取得
async function fetchPanoramaTile(
  panoId: string,
  zoom: number,
  x: number,
  y: number
): Promise<Buffer> {
  ensureDir(PANORAMA_DIR);
  const fileName = `${panoId}_z${zoom}_x${x}_y${y}.jpg`;
  const filePath = path.join(PANORAMA_DIR, fileName);

  // キャッシュがあれば使用
  if (fs.existsSync(filePath)) {
    return fs.readFileSync(filePath);
  }

  const url = `${TILE_BASE_URL}?cb_client=maps_sv.tactile&panoid=${panoId}&x=${x}&y=${y}&zoom=${zoom}&nbt=1&fover=2`;

  const response = await axios.get(url, {
    responseType: "arraybuffer",
    headers: {
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    },
    validateStatus: (status) => status < 500, // 4xx も受け入れて判定
  });

  if (response.status !== 200) {
    throw new Error(`Tile not available: ${response.status}`);
  }

  const buffer = Buffer.from(response.data);
  fs.writeFileSync(filePath, buffer);
  return buffer;
}

// パノラマタイルを結合してフルパノラマを作成（必要に応じて低いzoomレベルにフォールバック）
async function buildPanorama(panoId: string, requestedZoom: number): Promise<Buffer> {
  ensureDir(PANORAMA_DIR);

  // 利用可能なzoomレベルを試す（requestedZoomから下げていく）
  for (let zoom = requestedZoom; zoom >= 1; zoom--) {
    const cacheFile = path.join(PANORAMA_DIR, `${panoId}_full_z${zoom}.jpg`);

    // キャッシュがあれば使用
    if (fs.existsSync(cacheFile)) {
      if (zoom < requestedZoom) {
        console.log(`  パノラマキャッシュ使用 (zoom=${zoom}、要求より低い解像度)`);
      }
      return fs.readFileSync(cacheFile);
    }

    // zoom levelに応じたタイル数
    // zoom 0: 1x1, zoom 1: 2x1, zoom 2: 4x2, zoom 3: 8x4, zoom 4: 16x8, zoom 5: 32x16
    const tilesX = Math.pow(2, zoom);
    const tilesY = Math.pow(2, zoom - 1);
    const tileSize = 512;

    console.log(`  パノラマタイル取得中: zoom=${zoom}, tiles=${tilesX}x${tilesY}`);

    // 全タイルを取得
    const tiles: { x: number; y: number; buffer: Buffer }[] = [];
    let tilesFailed = false;

    for (let y = 0; y < tilesY && !tilesFailed; y++) {
      for (let x = 0; x < tilesX && !tilesFailed; x++) {
        try {
          const buffer = await fetchPanoramaTile(panoId, zoom, x, y);
          tiles.push({ x, y, buffer });
        } catch (error) {
          console.log(`  タイル取得失敗 (zoom=${zoom}, x=${x}, y=${y})、より低い解像度を試行...`);
          tilesFailed = true;
        }
      }
    }

    if (tilesFailed) {
      continue; // 次の低いzoomレベルを試す
    }

    // タイルを結合
    const width = tilesX * tileSize;
    const height = tilesY * tileSize;

    const compositeInputs = tiles.map((tile) => ({
      input: tile.buffer,
      left: tile.x * tileSize,
      top: tile.y * tileSize,
    }));

    const panorama = await sharp({
      create: {
        width,
        height,
        channels: 3,
        background: { r: 0, g: 0, b: 0 },
      },
    })
      .composite(compositeInputs)
      .jpeg({ quality: 90 })
      .toBuffer();

    fs.writeFileSync(cacheFile, panorama);
    return panorama;
  }

  throw new Error(`パノラマタイルを取得できませんでした: ${panoId}`);
}

// 等距円筒図法（equirectangular）からパースペクティブ投影で画像を切り出す
async function extractPerspective(
  panoramaBuffer: Buffer,
  heading: number, // 0-360度
  pitch: number, // -90 to 90度
  fov: number, // 視野角（度）
  outputWidth: number,
  outputHeight: number
): Promise<Buffer> {
  const panorama = sharp(panoramaBuffer);
  const metadata = await panorama.metadata();
  const panoWidth = metadata.width!;
  const panoHeight = metadata.height!;

  // ラジアンに変換
  const headingRad = (heading * Math.PI) / 180;
  const pitchRad = (pitch * Math.PI) / 180;
  const fovRad = (fov * Math.PI) / 180;

  // 出力画像の各ピクセルに対応するパノラマの座標を計算
  const outputPixels = new Uint8Array(outputWidth * outputHeight * 3);
  const rawPanorama = await panorama.raw().toBuffer();

  const halfFov = fovRad / 2;
  const aspectRatio = outputWidth / outputHeight;

  for (let outY = 0; outY < outputHeight; outY++) {
    for (let outX = 0; outX < outputWidth; outX++) {
      // 正規化された画面座標 (-1 to 1)
      const nx = (2 * outX) / outputWidth - 1;
      const ny = 1 - (2 * outY) / outputHeight;

      // 視線ベクトルを計算
      const x = nx * Math.tan(halfFov) * aspectRatio;
      const y = ny * Math.tan(halfFov);
      const z = 1;

      // 正規化
      const len = Math.sqrt(x * x + y * y + z * z);
      let vx = x / len;
      let vy = y / len;
      let vz = z / len;

      // pitch回転（X軸周り）
      const cosPitch = Math.cos(pitchRad);
      const sinPitch = Math.sin(pitchRad);
      const vy2 = vy * cosPitch - vz * sinPitch;
      const vz2 = vy * sinPitch + vz * cosPitch;
      vy = vy2;
      vz = vz2;

      // heading回転（Y軸周り）
      const cosHeading = Math.cos(headingRad);
      const sinHeading = Math.sin(headingRad);
      const vx2 = vx * cosHeading + vz * sinHeading;
      const vz3 = -vx * sinHeading + vz * cosHeading;
      vx = vx2;
      vz = vz3;

      // 球面座標に変換
      const theta = Math.atan2(vx, vz); // 経度 (-PI to PI)
      const phi = Math.asin(vy); // 緯度 (-PI/2 to PI/2)

      // パノラマ画像の座標に変換
      let panoX = ((theta + Math.PI) / (2 * Math.PI)) * panoWidth;
      let panoY = ((Math.PI / 2 - phi) / Math.PI) * panoHeight;

      // 範囲を制限
      panoX = Math.max(0, Math.min(panoWidth - 1, panoX));
      panoY = Math.max(0, Math.min(panoHeight - 1, panoY));

      // 最近傍補間でピクセル値を取得
      const srcX = Math.floor(panoX);
      const srcY = Math.floor(panoY);
      const srcIdx = (srcY * panoWidth + srcX) * 3;
      const dstIdx = (outY * outputWidth + outX) * 3;

      outputPixels[dstIdx] = rawPanorama[srcIdx];
      outputPixels[dstIdx + 1] = rawPanorama[srcIdx + 1];
      outputPixels[dstIdx + 2] = rawPanorama[srcIdx + 2];
    }
  }

  return sharp(outputPixels, {
    raw: { width: outputWidth, height: outputHeight, channels: 3 },
  })
    .jpeg({ quality: 90 })
    .toBuffer();
}

// 高解像度のStreet View画像を生成
// nextLat, nextLng を指定すると、その方向を向いた画像を取得する
export async function fetchHighResStreetViewImage(
  lat: number,
  lng: number,
  apiKey: string,
  nextLat: number,
  nextLng: number,
  outputWidth: number = 1280,
  outputHeight: number = 960,
  zoom: number = 3 // zoom 3 = 4096x2048, zoom 4 = 8192x4096
): Promise<string> {
  ensureDir(TMP_DIR);

  console.log(`高解像度Street View画像取得: (${lat}, ${lng}) → (${nextLat}, ${nextLng})`);

  // 実世界のheadingを計算
  const realWorldHeading = calculateHeading(lat, lng, nextLat, nextLng);
  const pitch = 0;
  const fov = 90;

  // キャッシュファイル名
  const cacheFile = path.join(
    TMP_DIR,
    `highres_${lat}_${lng}_h${Math.round(realWorldHeading)}_${outputWidth}x${outputHeight}.jpg`
  );

  if (fs.existsSync(cacheFile)) {
    console.log(`  ローカル高解像度画像を使用`);
    const buffer = fs.readFileSync(cacheFile);
    return buffer.toString("base64");
  }

  // pano_idを取得
  const panoId = await getPanoramaId(lat, lng, apiKey);
  console.log(`  pano_id: ${panoId}`);

  // パノラマのヘディングオフセットを取得
  const headingOffset = await getPanoramaHeadingOffset(panoId);
  console.log(`  heading offset: ${headingOffset.toFixed(2)}°`);

  // パノラマ座標系でのheadingを計算
  // パノラマのpanoX=0がheadingOffset度を向いているので、
  // 実世界のheadingからoffsetを引く
  const panoramaHeading = (realWorldHeading - headingOffset + 360) % 360;
  console.log(`  realWorldHeading: ${realWorldHeading.toFixed(2)}° → panoramaHeading: ${panoramaHeading.toFixed(2)}°`);

  // パノラマを構築
  const panorama = await buildPanorama(panoId, zoom);

  // パースペクティブ投影で切り出し
  const perspective = await extractPerspective(
    panorama,
    panoramaHeading,
    pitch,
    fov,
    outputWidth,
    outputHeight
  );

  // キャッシュに保存
  fs.writeFileSync(cacheFile, perspective);

  return perspective.toString("base64");
}

// 2点間の方位角を計算
export function calculateHeading(
  fromLat: number,
  fromLng: number,
  toLat: number,
  toLng: number
): number {
  const toRadians = (deg: number) => (deg * Math.PI) / 180;
  const toDegrees = (rad: number) => (rad * 180) / Math.PI;

  const lat1 = toRadians(fromLat);
  const lat2 = toRadians(toLat);
  const deltaLng = toRadians(toLng - fromLng);

  const x = Math.sin(deltaLng) * Math.cos(lat2);
  const y = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(deltaLng);

  let heading = toDegrees(Math.atan2(x, y));
  heading = (heading + 360) % 360;

  return heading;
}

File: claude-infer/python/panorama.py
"""Google Street Viewパノラマ画像の取得（Python版）"""

import math
import os
from io import BytesIO
from pathlib import Path

import numpy as np
import requests
from PIL import Image

from config import TMP_DIR

STREET_VIEW_METADATA_URL = "https://maps.googleapis.com/maps/api/streetview/metadata"
TILE_BASE_URL = "https://streetviewpixels-pa.googleapis.com/v1/tile"
PANORAMA_DIR = TMP_DIR / "panorama"


def get_panorama_id(lat: float, lng: float, api_key: str) -> str:
    """Street Viewのpanorama IDを取得"""
    response = requests.get(
        STREET_VIEW_METADATA_URL,
        params={
            "location": f"{lat},{lng}",
            "key": api_key,
            "source": "outdoor",
        }
    )
    data = response.json()

    if data.get("status") != "OK" or not data.get("pano_id"):
        raise Exception(f"Street View not available: {data.get('status')}")

    copyright_text = data.get("copyright", "")
    if copyright_text and "Google" not in copyright_text:
        raise Exception(f"User-contributed image skipped: {copyright_text}")

    return data["pano_id"]


def get_panorama_heading_offset(pano_id: str) -> float:
    """パノラマのヘディングオフセットを取得"""
    try:
        url = f"https://www.google.com/maps/photometa/v1?authuser=0&hl=ja&gl=jp&pb=!1m4!1smaps_sv.tactile!11m2!2m1!1b1!2m2!1sja!2sjp!3m3!1m2!1e2!2s{pano_id}!4m57!1e1!1e2!1e3!1e4!1e5!1e6!1e8!1e12!2m1!1e1!4m1!1i48!5m1!1e1!5m1!1e2!6m1!1e1!6m1!1e2!9m36!1m3!1e2!2b1!3e2!1m3!1e2!2b0!3e3!1m3!1e3!2b1!3e2!1m3!1e3!2b0!3e3!1m3!1e8!2b0!3e3!1m3!1e1!2b0!3e3!1m3!1e4!2b0!3e3!1m3!1e10!2b1!3e2!1m3!1e10!2b0!3e3"

        response = requests.get(url, headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        })

        text = response.text
        json_str = text.replace(")]}'\\n", "").replace(")]}\\'\\n", "")
        if json_str.startswith(")]}'"):
            json_str = json_str[4:]

        import json
        data = json.loads(json_str)

        heading_data = data[1][0][5][0][1][2] if data else None
        if heading_data and isinstance(heading_data, list) and len(heading_data) >= 1:
            heading = heading_data[0]
            if isinstance(heading, (int, float)):
                return float(heading)

        return 0.0
    except Exception:
        return 0.0


def fetch_panorama_tile(pano_id: str, zoom: int, x: int, y: int) -> bytes:
    """パノラマタイルを取得"""
    PANORAMA_DIR.mkdir(parents=True, exist_ok=True)

    cache_file = PANORAMA_DIR / f"{pano_id}_z{zoom}_x{x}_y{y}.jpg"
    if cache_file.exists():
        return cache_file.read_bytes()

    url = f"{TILE_BASE_URL}?cb_client=maps_sv.tactile&panoid={pano_id}&x={x}&y={y}&zoom={zoom}&nbt=1&fover=2"

    response = requests.get(url, headers={
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
    })

    if response.status_code != 200:
        raise Exception(f"Tile not available: {response.status_code}")

    content_type = response.headers.get("Content-Type", "")
    if not content_type.startswith("image/"):
        raise Exception(f"Invalid response: expected image, got {content_type}")

    if len(response.content) < 1000:
        raise Exception(f"Invalid tile data: too small ({len(response.content)} bytes)")

    cache_file.write_bytes(response.content)
    return response.content


def build_panorama(pano_id: str, zoom: int = 3) -> Image.Image:
    """パノラマタイルを結合してフルパノラマを作成"""
    PANORAMA_DIR.mkdir(parents=True, exist_ok=True)

    cache_file = PANORAMA_DIR / f"{pano_id}_full_z{zoom}.jpg"
    if cache_file.exists():
        return Image.open(cache_file)

    # zoom levelに応じたタイル数
    tiles_x = 2 ** zoom
    tiles_y = 2 ** (zoom - 1)
    tile_size = 512

    width = tiles_x * tile_size
    height = tiles_y * tile_size

    panorama = Image.new("RGB", (width, height), (0, 0, 0))

    for y in range(tiles_y):
        for x in range(tiles_x):
            try:
                tile_data = fetch_panorama_tile(pano_id, zoom, x, y)
                tile_image = Image.open(BytesIO(tile_data))
                panorama.paste(tile_image, (x * tile_size, y * tile_size))
            except Exception as e:
                print(f"  タイル取得失敗: zoom={zoom}, x={x}, y={y}: {e}")
                raise

    panorama.save(cache_file, "JPEG", quality=90)
    return panorama


def extract_perspective(
    panorama: Image.Image,
    heading: float,
    pitch: float,
    fov: float,
    output_width: int,
    output_height: int
) -> Image.Image:
    """等距円筒図法からパースペクティブ投影で画像を切り出す"""
    pano_width, pano_height = panorama.size
    pano_array = np.array(panorama)

    heading_rad = math.radians(heading)
    pitch_rad = math.radians(pitch)
    fov_rad = math.radians(fov)

    output_array = np.zeros((output_height, output_width, 3), dtype=np.uint8)

    half_fov = fov_rad / 2
    aspect_ratio = output_width / output_height

    for out_y in range(output_height):
        for out_x in range(output_width):
            # 正規化された画面座標 (-1 to 1)
            nx = (2 * out_x) / output_width - 1
            ny = 1 - (2 * out_y) / output_height

            # 視線ベクトルを計算
            x = nx * math.tan(half_fov) * aspect_ratio
            y = ny * math.tan(half_fov)
            z = 1

            # 正規化
            length = math.sqrt(x * x + y * y + z * z)
            vx, vy, vz = x / length, y / length, z / length

            # pitch回転（X軸周り）
            cos_pitch = math.cos(pitch_rad)
            sin_pitch = math.sin(pitch_rad)
            vy2 = vy * cos_pitch - vz * sin_pitch
            vz2 = vy * sin_pitch + vz * cos_pitch
            vy, vz = vy2, vz2

            # heading回転（Y軸周り）
            cos_heading = math.cos(heading_rad)
            sin_heading = math.sin(heading_rad)
            vx2 = vx * cos_heading + vz * sin_heading
            vz3 = -vx * sin_heading + vz * cos_heading
            vx, vz = vx2, vz3

            # 球面座標に変換
            theta = math.atan2(vx, vz)  # 経度 (-PI to PI)
            phi = math.asin(max(-1, min(1, vy)))  # 緯度 (-PI/2 to PI/2)

            # パノラマ画像の座標に変換
            pano_x = ((theta + math.pi) / (2 * math.pi)) * pano_width
            pano_y = ((math.pi / 2 - phi) / math.pi) * pano_height

            # 範囲を制限
            pano_x = max(0, min(pano_width - 1, pano_x))
            pano_y = max(0, min(pano_height - 1, pano_y))

            # 最近傍補間でピクセル値を取得
            src_x = int(pano_x)
            src_y = int(pano_y)
            output_array[out_y, out_x] = pano_array[src_y, src_x]

    return Image.fromarray(output_array)


def calculate_heading(from_lat: float, from_lng: float, to_lat: float, to_lng: float) -> float:
    """2点間の方位角を計算"""
    lat1 = math.radians(from_lat)
    lat2 = math.radians(to_lat)
    delta_lng = math.radians(to_lng - from_lng)

    x = math.sin(delta_lng) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(delta_lng)

    heading = math.degrees(math.atan2(x, y))
    heading = (heading + 360) % 360

    return heading


def fetch_street_view_image(
    lat: float,
    lng: float,
    api_key: str,
    next_lat: float,
    next_lng: float,
    output_width: int = 1280,
    output_height: int = 960,
    zoom: int = 3
) -> str:
    """高解像度のStreet View画像を取得（Base64）"""
    import base64

    TMP_DIR.mkdir(exist_ok=True)

    # 実世界のheadingを計算
    real_world_heading = calculate_heading(lat, lng, next_lat, next_lng)
    pitch = 0
    fov = 90

    # キャッシュファイル名
    cache_file = TMP_DIR / f"highres_{lat}_{lng}_h{round(real_world_heading)}_{output_width}x{output_height}.jpg"

    if cache_file.exists():
        with open(cache_file, "rb") as f:
            return base64.b64encode(f.read()).decode()

    # pano_idを取得
    pano_id = get_panorama_id(lat, lng, api_key)

    # パノラマのヘディングオフセットを取得
    heading_offset = get_panorama_heading_offset(pano_id)

    # パノラマ座標系でのheadingを計算
    panorama_heading = (real_world_heading - heading_offset + 360) % 360

    # パノラマを構築
    panorama = build_panorama(pano_id, zoom)

    # パースペクティブ投影で切り出し
    perspective = extract_perspective(
        panorama,
        panorama_heading,
        pitch,
        fov,
        output_width,
        output_height
    )

    # キャッシュに保存
    perspective.save(cache_file, "JPEG", quality=90)

    with open(cache_file, "rb") as f:
        return base64.b64encode(f.read()).decode()


def download_image(
    lat: float,
    lng: float,
    api_key: str,
    next_lat: float,
    next_lng: float,
    output_width: int = 1280,
    output_height: int = 960,
) -> Path:
    """画像をダウンロードしてパスを返す"""
    real_world_heading = calculate_heading(lat, lng, next_lat, next_lng)
    cache_file = TMP_DIR / f"highres_{lat}_{lng}_h{round(real_world_heading)}_{output_width}x{output_height}.jpg"

    if cache_file.exists():
        return cache_file

    fetch_street_view_image(
        lat, lng, api_key, next_lat, next_lng,
        output_width, output_height
    )

    return cache_file
